core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -215,14 +227,93 @@ public boolean isAvailable() {
       return file.exists();
    }
 
+   private boolean isCorruptData() throws IOException {
+      // ISPN-13128 Corrupt migration data can only be created with default marshaller
+      if (ctx.getGlobalConfiguration().serialization().marshaller() != null)
+         return false;
+
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      long pos = filePos;
+      buf = readChannel(buf, pos, KEY_POS_LATEST);
+
+      int remaining = buf.remaining();
+      if (buf.remaining() > 0) {
+         // Remaining == KEY_POS_LATEST means the file was empty and nothing read
+         return remaining != KEY_POS_LATEST;
+      }
+      buf.flip();
+      FileEntry fe = new FileEntry(pos, buf);
+      channel.position(filePos);
+      // Explicitly calculate the correct size as incorrect value stored in fe.size
+      return fe.size != fe.actualSize();
+   }
+
+   private void migrateCorruptData() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      long oldFilePos = MAGIC_LATEST.length;
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            oldFilePos += KEY_POS_LATEST;
+
+            // Read old entry content and then write
+            ByRef.Long offset = new ByRef.Long(oldFilePos);
+            K key = unmarshallObject(buf, offset, fe.keyLen);
+            V value = unmarshallObject(buf, offset, fe.dataLen);
+
+            int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+            Metadata metadata = unmarshallObject(buf, offset, metaLen);
+            // internalMetadata was ALWAYS lost - so don't read it
+
+            oldFilePos = offset.get();
+
+            // LastUsed and Created timestamps were lost, set creation time to current time if fe.expiryTime has been set
+            long currentTS = fe.expiryTime > 0 ? timeService.wallClockTime() : -1;
+            MarshallableEntry<? extends K, ? extends V> me = (MarshallableEntry<? extends K, ? extends V>) ctx.getMarshallableEntryFactory().create(key, value, metadata, null, currentTS, currentTS);
```

Comment:
I'm not sure about using `null` for internal metadata. I know we will have issues with xsite because the versions are lost but I'm not sure if `null` is ok for optimistic transactions :/

Reply:
Right, I forgot about initializing these values. Based upon previous conversations, WDYT to the following approach:

Optimistic tx -> `PrivateMetadata` with version == 1

XSite:

- Attempt to recover latest version from global state when shutdown
- Advise performing xsite state transfer once corrupt data migrated/

Reply:
looks good.